activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb

```diff
@@ -61,6 +61,19 @@ def add_column_options!(sql, options)
             if options[:collation]
               sql << " COLLATE \"#{options[:collation]}\""
             end
+
+            if as = options[:as]
+              sql << " GENERATED ALWAYS AS (#{as})"
+
+              if options[:stored]
+                sql << " STORED"
+              else
+                raise ArgumentError, <<~MSG
```

Comment:
This is the only place where I disagree with some of the given feedback regarding future proofing (e.g. https://github.com/rails/rails/pull/39368#discussion_r499937396). In PostgreSQL 14 there is still no `VIRTUAL` support planed.
In my opinion it's better to provide a precise error message rather than adding support for something that may not happen anytime soon (and maybe even create incorrect SQL statements in the process).

Note: This only affects migration support. Reading any type of virtual column is supported.